File: backend/app/routers/inventory.py
# app/routers/inventory.py
from typing import List

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlmodel import Session, select

from ..db import get_session
from ..models import InventoryItem, InventoryStock, StockMovement, MovementReason

router = APIRouter(prefix="/inventory", tags=["inventory"])


class StockUpsertPayload(BaseModel):
    site_id: int
    item_id: int
    quantity: int
    min_level_override: int | None = None


class StockMovePayload(BaseModel):
    delta: int
    reason: MovementReason = MovementReason.usage
    reference: str | None = None
    author: str | None = None


# ----- Items -----
@router.get("/items", response_model=List[InventoryItem])
def list_items(session: Session = Depends(get_session)):
    return session.exec(select(InventoryItem).order_by(InventoryItem.name)).all()


@router.post("/items", response_model=InventoryItem)
def create_item(item: InventoryItem, session: Session = Depends(get_session)):
    session.add(item)
    session.commit()
    session.refresh(item)
    return item


@router.put("/items/{item_id}", response_model=InventoryItem)
def update_item(
    item_id: int,
    data: InventoryItem,
    session: Session = Depends(get_session),
):
    it = session.get(InventoryItem, item_id)
    if not it:
        raise HTTPException(404, "Item not found")

    for k, v in data.model_dump(exclude_unset=True).items():
        setattr(it, k, v)

    session.add(it)
    session.commit()
    session.refresh(it)
    return it


@router.delete("/items/{item_id}")
def delete_item(item_id: int, session: Session = Depends(get_session)):
    it = session.get(InventoryItem, item_id)
    if not it:
        raise HTTPException(404, "Item not found")
    session.delete(it)
    session.commit()
    return {"ok": True}


# ----- Stock -----
@router.get("/stock", response_model=List[InventoryStock])
def list_stock(site_id: int, session: Session = Depends(get_session)):
    return session.exec(
        select(InventoryStock).where(InventoryStock.site_id == site_id)
    ).all()


@router.post("/stock/upsert", response_model=InventoryStock)
def upsert_stock(payload: StockUpsertPayload, session: Session = Depends(get_session)):
    row = session.exec(
        select(InventoryStock).where(
            (InventoryStock.site_id == payload.site_id)
            & (InventoryStock.item_id == payload.item_id)
        )
    ).first()

    if row:
        row.quantity = payload.quantity
        if payload.min_level_override is not None:
            row.min_level_override = payload.min_level_override
        session.add(row)
        session.commit()
        session.refresh(row)
        return row

    new_row = InventoryStock(
        site_id=payload.site_id,
        item_id=payload.item_id,
        quantity=payload.quantity,
        min_level_override=payload.min_level_override,
    )
    session.add(new_row)
    session.commit()
    session.refresh(new_row)
    return new_row


@router.post("/stock/{stock_id}/move", response_model=StockMovement)
def move_stock(
    stock_id: int,
    payload: StockMovePayload,
    session: Session = Depends(get_session),
):
    st = session.get(InventoryStock, stock_id)
    if not st:
        raise HTTPException(404, "Stock not found")

    st.quantity = (st.quantity or 0) + payload.delta

    mv = StockMovement(
        stock_id=stock_id,
        delta_qty=payload.delta,
        reason=payload.reason,
        reference=payload.reference,
        author=payload.author,
    )

    session.add(st)
    session.add(mv)
    session.commit()
    session.refresh(mv)
    return mv

File: frontend/src/services/inventory.ts
// src/services/inventory.ts
import api from "../lib/api";

export type Item = {
  id: number;
  sku: string;
  name: string;
  category?: string | null;
  uom: string;
  notes?: string | null;
};

export type Stock = {
  id: number;
  site_id: number;
  item_id: number;
  quantity: number;
  min_level_override?: number | null;
  updated_at: string;
};

export type MovementReason =
  | "usage"
  | "delivery"
  | "adjustment"
  | "transfer";

// -------- Items --------

export async function listItems(): Promise<Item[]> {
  const res = await api.get<Item[]>("/inventory/items");
  return res.data;
}

export async function createItem(payload: Omit<Item, "id">): Promise<Item> {
  const res = await api.post<Item>("/inventory/items", payload);
  return res.data;
}

export async function updateItem(
  id: number,
  payload: Partial<Omit<Item, "id">>
): Promise<Item> {
  const res = await api.put<Item>(`/inventory/items/${id}`, payload);
  return res.data;
}

export async function deleteItem(id: number): Promise<void> {
  await api.delete(`/inventory/items/${id}`);
}

// -------- Sites --------

export type SiteRef = { id: number; name: string };

export async function listInventorySites(): Promise<SiteRef[]> {
  const res = await api.get<SiteRef[]>("/sites/");
  return res.data;
}

// -------- Stock --------

export async function getStockForSite(siteId: number): Promise<Stock[]> {
  const res = await api.get<Stock[]>(`/inventory/stock?site_id=${siteId}`);
  return res.data;
}

export async function upsertStock(
  siteId: number,
  itemId: number,
  qty: number
): Promise<void> {
  await api.post("/inventory/stock/upsert", {
    site_id: siteId,
    item_id: itemId,
    quantity: qty,
  });
}

export async function moveStock(
  stockId: number,
  delta: number,
  reason: MovementReason,
  reference: string
): Promise<void> {
  await api.post(`/inventory/stock/${stockId}/move`, {
    delta,
    reason,
    reference,
    author: "web",
  });
}
